### DISCUSSION TITLE
Example for Teams link unfurling

### DISCUSSION THREAD

**Discussion by yukun-dong** [microsoft/teams-sdk#138] [Q&A] [answered]
**Describe your user scenario**

> I want to use this SDK to build a link unfurling app. I see you already provide APIs for `queryLink` and `anonymousQueryLink` but I don't know how to use them. In specific, I don't know what should be passed into `commandId` since link unfurling doesn't have a commandId in manifest. Therefore, It would be great if you can provide a sample of this.
[MessageExtensions.ts L260](https://github.com/microsoft/botbuilder-m365/blob/ff963f8ed01c9d1d372b4f315afbcd671a2e5bae/js/packages/botbuilder-m365/src/MessageExtensions.ts#L260)
[MessageExtensions.ts L39](https://github.com/microsoft/botbuilder-m365/blob/ff963f8ed01c9d1d372b4f315afbcd671a2e5bae/js/packages/botbuilder-m365/src/MessageExtensions.ts#L39)

**Describe the solution you'd like**

> Provide a js sample for link unfurling.

**Describe alternatives you've considered**

> Provide documents on how to use `queryLink` and `anonymousQueryLink`.

**Comment by corinagum** [comment, marked as answer]
Hi @yukun-dong,

1. First, may I ask if you have seen this [documentation on Link unfurling](https://learn.microsoft.com/en-us/microsoftteams/platform/messaging-extensions/how-to/link-unfurling?tabs=desktop%2Cjavascript%2Cadvantages)?  There is some setup in the dev portal that you may or may not already be aware of. The documentation includes how to:
    - Enable message extensions on your bot when manual unfurling is not necessary
    - Update app manifest `composeExtensions`
    - See [invoke invoke handler documentation](https://learn.microsoft.com/en-us/microsoftteams/platform/messaging-extensions/how-to/link-unfurling?tabs=desktop%2Cjavascript%2Cadvantages#handle-the-composeextensionquerylink-invoke)
3. Similar to the [search extension sample](https://github.com/microsoft/teams-ai/js/samples/02.messageExtensions.a.searchCommand/src/index.ts) you'll set up your queryLink payload to return the card you want to unfurl based off of the query link

```ts
app.messageExtensions.queryLink('unfurlCmd', async (context, state): MessageExtensionActionResponse => {
   // get url from activity (or possibly parse url from text)
   const queryLink = context.activity.value 
   // TODO: validate queryLink is valid URL
   // TODO: potentially do API call with valid URL to get details (e.g. image url, title, etc) e.g.:
   const response = await axios.get(queryLink.ApiCall());

   const card = AdaptiveCardPayload;    // AdaptiveCardPayload can be a pre-formatted Adaptive Card json file. 
   card.imageUrl = response.image.imgUrl;
   card.subtitle = response.image.altText;
   return card;
});

// result: new activity with Adaptive Card attachment is sent
```

Since there's plenty of documentation already in existence, I think that a sample will be unnecessary. :)

**Comment by yukun-dong** [comment]
Hi @corinagum , thanks for the reply.
Actually I want to figure out what's the first parameter that should be passed to `app.messageExtensions.queryLink()`, because link unfurling is different to action and search, it doesn't have a commandId.

**Comment by corinagum** [comment]
@yukun-dong The command id can be whatever you (the developer) wants it to be. The first parameter is the string or RegExp pattern you will use to call it within the bot, as needed. Since the code for this sends the activity, you will want to call it explicitly.

```ts
    /**
     * Registers a handler that implements a Link Unfurling based Message Extension.
     * @param commandId ID of the command(s) to register the handler for.
     * @param handler Function to call when the command is received.
     * @param handler.context Context for the current turn of conversation with the user.
     * @param handler.state Current state of the turn.
     * @param handler.url The URL that should be unfurled.
     * @returns The application for chaining purposes.
     */
    public queryLink(
```
